### ISSUE TITLE
Segmentation fault in ComputeMatches

### ISSUE THREAD

**Issue by rdiazgar** [openMVG/openMVG#278]
I'm finding impossible to run a large feature match in an 6400 image set of my own. ComputeMatches always returns a segmentation fault that seems random in the source error, but always appears in the same function call. The console spits out the following output:

 - GEOMETRIC FILTERING - 

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
**** glibc detected *** software/SfM/openMVG_main_computeMatches: munmap_chunk(): invalid pointer: 0x00002b8c08002880 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3ddee75e66]
software/SfM/openMVG_main_computeMatches(_ZN7openMVG6robust8ACRANSACINS0_24ACKernelAdaptorEssentialINS_9essential6kernel15EssentialKernelINS4_15
6MatrixIdLi3ELi3ELi0ELi3ELi3EEEEENS8_21EpipolarDistanceErrorENS_13UnnormalizerTESC_EEEESt4pairIddERKT_RSt6vectorImSaImEEmPNSJ_5ModelEdb+0x12e7)[
software/SfM/openMVG_main_computeMatches[0x42d3a1]
/pkg/gcc/4.8.2/lib64/libgomp.so.1(+0x838a)[0x2b8ae3c5038a]
/lib64/libpthread.so.0[0x3ddfa079d1]
/lib64/libc.so.6(clone+0x6d)[0x3ddeee88fd]
======= Memory map: ========
00400000-00566000 r-xp 00000000 00:1d 136352347                          /extra/titansc0/rdiazgar/software/openMVG/build/software/SfM/openMVG_ma
00766000-00767000 rw-p 00166000 00:1d 136352347                          /extra/titansc0/rdiazgar/software/openMVG/build/software/SfM/openMVG_ma
00767000-00768000 rw-p 00000000 00:00 0 
[Error trace continues longer ...]

Trying to run the process several times ends always with the seg fault. Sometimes the error is "munmap_chunk()" like above, some other times it's a "double free()" issue, or "corrupted double-linked list", etc. It always seems to happen around the UnnormalizerT function. It looks to me like some sort of library issue because this error happens on a CentOS 6.4 48-core server, but a simple 8-core desktop with ubuntu 12.04 up to date works fine.

Ever encountered a similar problem? I'm happy to provide other useful output like ldd's to find the solution.

**Comment by pmoulon** [comment]
Hi,
So if I resume:
The error only appear when you run the process on your CentOs.
But when you launch the process with the same data on your ubuntu the process run fine.

I have never encountered this problem.
Could you tell me if you are using the master or develop branch ?
Have you tried to run with no openMP, I mean set only one thread (to see if the error comes from a parallelism issue).
I mean run:
$ set OMP_NUM_THREADS=1
$ openMVG_main_computeMatches ...

**Comment by FlachyJoe** [comment]
Hi,
I also get this error after an update of my system but it fixes with git 
pull and recompilation.

**Comment by rdiazgar** [comment]
Hi,

I've tried several things so far (always using the master branch):

- If the error is in UnnormalizerT, could it be some errors in the Eigen/Lapack/Blas libraries when denormalizing a model? I tried compiling with and without Lapack/Blas and makes no difference.

- OpenMP/threading issues: 

I have little experience using OpenMP, but after googling for a while this seems the behaviour of a race condition problem. Could it be that the "GeometricFilterT filter = geometricFilter;" in the threaded for loop is not being copied safely somehow? I tried simple approaches like creating this copy before the for loop and the adding the "firstprivate(filter)" directive with the same results.

I tried running with 1 thread, although your "set OMP_NUM_THREADS=1" did not have effect and ran on multiple CPU's. Instead, I recompiled using the num_threads(1) instead of schedule(dynamic,1). Same segfault results though. Other schedule directives like 'static' do not work either.

If it helps, here's the output of the process under valgrind:

 - GEOMETRIC FILTERING - 

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
vex amd64->IR: unhandled instruction bytes: 0xC4 0xE3 0x69 0x6F 0x5C 0x24 0x40 0x30
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=1 VEX.L=0 VEX.nVVVV=0x2 ESC=0F3A
vex amd64->IR:   PFX.66=1 PFX.F2=0 PFX.F3=0
==42733== valgrind: Unrecognised instruction at address 0x42d061.
==42733==    at 0x42D061: void ImageCollectionGeometricFilter<openMVG::SIOPointFeature>::Filter<openMVG::GeometricFilter_EMatrix_AC>(openMVG::GeometricFilter_EMatrix_AC const&, std::map<std::pair<unsigned 
long, unsigned long>, std::vector<openMVG::matching::IndMatch, std::allocator<openMVG::matching::IndMatch> >, std::less<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<std::pair<unsigne
d long, unsigned long> const, std::vector<openMVG::matching::IndMatch, std::allocator<openMVG::matching::IndMatch> > > > >&, std::map<std::pair<unsigned long, unsigned long>, std::vector<openMVG::matching:
:IndMatch, std::allocator<openMVG::matching::IndMatch> >, std::less<std::pair<unsigned long, unsigned long> >, std::allocator<std::pair<std::pair<unsigned long, unsigned long> const, std::vector<openMVG::m
atching::IndMatch, std::allocator<openMVG::matching::IndMatch> > > > >&, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > > const&) const [clone
 .omp_fn.6] (in /extra/titansc0/rdiazgar/software/openMVG/build/software/SfM/openMVG_main_computeMatches)
==42733==    by 0x4C17389: gomp_thread_start (team.c:115)
==42733==    by 0x3DDFA079D0: start_thread (pthread_create.c:301)
==42733==    by 0x3DDEEE88FC: clone (clone.S:115)
[error trace continues longer...]

Thanks
Raúl

**Comment by pmoulon** [comment]
Thanks Raul,
I think that the problem is due to multithreading issue or to memory issue.
Using a fresh rebuild can perhaps helps.

Soon you will be able to try with the 0.8 version and see if the bug is still here.

**Comment by rdiazgar** [comment]
Hi,

I'm not a 100% sure yet but the problem might be solved. The CentOS server has several versions of gcc installed, defaulting in 4.4.X. I switched to the newest one available, 4.8.X. Still, cmake was using 4.4 for some reason so I had to force cmake to use 4.8 with the DCMAKE_CXX_COMPILER and DCMAKE_C_COMPILER flags to re-build everything. 

ComputeMatches is now running smooth using the 48 cores of the server and so far it has not crashed. As I mentioned in my first post, the image set comprises 6400 images, so it'll take a while. Perhaps we can wait until tomorrow to see if it has not crashed and then close this issue?

As a side note, maybe gcc 4.4 is not a good compiler version for thread safety? My ubuntu desktop uses gcc 4.6 and runs good with it so this makes sense.

Regards
Raúl
